db/migrate/20240909151818_migrate_legacy_default_theme_to_light_theme.rb

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+class MigrateLegacyDefaultThemeToLightTheme < ActiveRecord::Migration[7.1]
+  class MigrationUserPreference < ActiveRecord::Base
+    self.table_name = "user_preferences"
+  end
+
+  # Migrate legacy `default` theme to `light` theme
+  #
+  # Moving from:
+  # {
+  #   "settings": {
+  #     "theme": "default"
+  #   }
+  # }
+  #
+  # To:
+  # {
+  #   "settings": {
+  #     "theme": "light" # current default theme
+  #   }
+  # }
+  def up
+    MigrationUserPreference
+      .where(settings: { theme: "default" })
+      .update_all(settings: { theme: "light" })
+  end
```

Comment:
This will not find settings if anything but theme is set, which is probably the main case.
I assume either walking using `find_each` and filtering/updating in ruby, or updating in SQL, finding using `jsonb_extract_path` and updating using `jsonb_set` should be done. SQL way should be faster, so less chances of overwriting settings if user is changing them in parallel.

Reply:
Ah interesting, didn't think of this. Now looking at the SQL that's spit out, you're right. Thought it was a partial match and inferred that it's a jsonb column. Nice catch

Reply:
I actually can't replicate the above. I've ran this migration against a couple of different scenarios with different users and it does find and update the right columns. But I understand the plain SQL way is probably a bit faster.